### ISSUE TITLE
Difficult to use some filters, when user has no data permissions [FE - Filter widget stops working if API endpoint returns 403]

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#15119]
**Describe the bug**
When a user does not have Data access, the filter dropdowns ("A list of all values" in Data Model) shows text input, but the user needs to add a filter value manually and then click elsewhere (blur) and then focus the filter input again, only now it's possible to filter the desired value.

**To Reproduce**
1. Admin > People > create user "U1"
2. Admin > Permissions > revoke Data access for Sample Dataset, and allow collection view/curate on Our Analytics
3. Simple question > Sample Dataset > Products - save as "Q1"
4. Add "Q1" to dashboard "D1" and add a filter Categories and connect to "Q1" Product.Category
5. Login as "U1" and go to dashboard "D1" - there are log warnings like this:
```
2021-03-10 15:08:56,911 DEBUG middleware.log :: GET /api/dashboard/:id 200 31.8 ms (10 DB calls) App DB connections: 0/13 Jetty threads: 4/50 (1 idle, 0 queued) (99 total active threads) Queries in flight: 0 (0 queued)
2021-03-10 15:08:57,097 WARN middleware.log :: GET /api/table/:id/query_metadata 403 6.4 ms (3 DB calls) 
"You don't have permissions to do that."
2021-03-10 15:08:57,274 WARN middleware.log :: GET /api/field/:id/values 403 8.4 ms (4 DB calls) 
"You don't have permissions to do that."
2021-03-10 15:08:57,284 WARN middleware.log :: GET /api/field/:id 403 22.3 ms (7 DB calls) 
"You don't have permissions to do that."
```
6. Click the filter - it does not show a dropdown - manually type `Gizmo`, but the "Add filter" button does not get enabled. Clicking somewhere outside of the filter popover (blur) and then clicking the filter again will show the previously entered value and now the "Add filter" button is enabled and the filter can be applied. Similar thing happens when trying to remove a filter value.
![Screencast-2021-03-10_15 14 01](https://user-images.githubusercontent.com/1447303/110642905-c9d15380-81b3-11eb-9d6e-af172b9f24aa.gif)
7. When inputting a value and pressing enter, then the input field is cleared and it looks like the filter isn't working. Clicking somewhere outside of the filter popover and then clicking the filter again will show the value and "Add filter" can be clicked.

Chrome throws the following error in the console, when clicking the filter before inputting a value:
```
app-main.bundle.js?6af835044d5f3f6749b7:5 Uncaught (in promise) TypeError: Cannot read property 'filter' of undefined
    at t._updateFilteredValues (app-main.bundle.js?6af835044d5f3f6749b7:5)
    at t.value (app-main.bundle.js?6af835044d5f3f6749b7:5)
    at c.updateComponent (vendor.bundle.js?6af835044d5f3f6749b7:485)
    at c.receiveComponent (vendor.bundle.js?6af835044d5f3f6749b7:485)
    at Object.receiveComponent (vendor.bundle.js?6af835044d5f3f6749b7:92)
    at Object.updateChildren (vendor.bundle.js?6af835044d5f3f6749b7:485)
    at Z._reconcilerUpdateChildren (vendor.bundle.js?6af835044d5f3f6749b7:485)
    at Z._updateChildren (vendor.bundle.js?6af835044d5f3f6749b7:485)
    at Z.updateChildren (vendor.bundle.js?6af835044d5f3f6749b7:485)
    at Z._updateDOMChildren (vendor.bundle.js?6af835044d5f3f6749b7:485)
```

**Information about your Metabase Installation:**
Regression since 0.37.0 - worked in 0.36.8 - tested 0.36.8 thru 0.38.1

**Additional context**
Related to #14309 #13832 #13537 #12654 (historic #8472)

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by willnaoosmith** [comment]
Same here!

**Comment by camsaul** [comment]
See #13537

**Comment by camsaul** [comment]
The various metadata endpoints returning 403 are by design. They're metadata for _creating_ new queries. We don't need to be calling them to _run_ queries. They've always required data read perms to access. The fix for those is that FE client should stop calling those endpoints when it's just _running_ a query. Either way, it's a separate issue and not related to the parameters endpoint not working

**Comment by ariya** [comment]
There seems to be a different issue at play here.

When I run v36, the dashboard front-end code (via metabase lib) still asked for `/api/table/:id/query_metadata` and the back-end also responsed with 403. However, this 403 response did not prevent the functionality of the dashboard at all.

Thus, I'm trying to find out, at which point after v36, such a 403 response broke the dashboard.

**Comment by ariya** [comment]
So far I managed to bisect the commits during v37 and thus, I narrowed it down to the following commit: 968a30766.
Running the above commit (or later) demonstrates the problem described in the issue, while running a commit before (05a3de54a) exhibits the same behavior as v36.

Since this problematic commit is a merge commit, my next step is trace both its parent commits to narrow down the regression further.

**Comment by ariya** [comment]
Apparently, this regression was introducing during the addition of chain filters (commit c4353a1f8). However, two changes in PR #13481 and PR ##15943 seem to fix this issue.

The dashboard code still hits `/api/fields/4` and `/api/fields/4/values` (both returning 403), but the field widget now obtain its values correctly through DashboardApi.parameterValues (e.g. `/api/dashboard/1/param/1ef75358/values`) instead (returning 200). @flamber could you also please confirm?

Note that the widget still says _Text_ instead of _Category_ but likely that's a separate issue (if it's an issue at all).

**Comment by flamber** [comment]
@ariya Yeah, it's still hitting the /api/field/:id and /api/field/:id/values, which 403 (or 401 via embedding), but it mostly work with #15943 - though it borks in one case, which is likely caused, when there's more than 300 distinct values in a field:
1. Admin > Data Model > Sample Dataset > Reviews > Reviewer > set as Category and "A list of all values"
Do the same for Rating field
2. Admin > People > create a user "U1"
3. Admin > Permissions > Sample Dataset > revoke permissions for Reviews
4. Simple question > Sample Dataset > Reviews - save as "Q1" and add to dashboard
5. On dashboard, create a category filter and link to "Q1".Reviewer and another category for "Q1".Rating - save dashboard - test that it works as admin
6. Login as "U1" - go to dashboard - clicking the Rating filter works, but clicking Reviewer filter will make it flicker and completely break (not clickable anymore)
This is basically the same behavior as the Embed Preview also does for all dropdown filters.

**Comment by ariya** [comment]
> Yeah, it's still hitting the /api/field/:id and /api/field/:id/values

Those endpoints will not be needed anymore with PR #16056.
And thanks for another test scenario, will investigate that as well.

**Comment by ariya** [comment]
@flamber should we close this one and track the excessive value requests in #16103?

**Comment by flamber** [comment]
@ariya We can create a separate issue, since this has become a P1 since 39, and #16103 is just a P3.

**Comment by flamber** [comment]
Closing as fixed by #15943 for most fields, but still fails for fields with 300+ distinct values, which is now P1 #16112
